### PR TITLE
First iteration of the frontend

### PR DESCRIPTION
**Pull request by fadeev**
This is the first iteration of the frontend. The code hasn't been refactored, so some things are not pretty. The goal right now is to make sure it works and ship it, then iterate on making it better.

## Portfolio and Token Transfer

Screenshot 2023-11-20 at 12 10 17

Features:

* Balances (native, supported ERC-20, ZRC-20, WZETA, BTC) across all connected chains
* Cross-chain fees are displayed (where applicable)
* Set custom destination address (where applicable)
* Quoter for cross-chain swaps
* Multi-functional swap component that supports token transfers and cross-chain swaps:
  - [x] crossChainZeta
  - [x] depositZRC20
  - [x] depositBTC
  - [x] transferNativeEVM
  - [x] withdrawBTC
  - [x] transferERC20EVM
  - [x] wrapZeta
  - [x] unwrapZeta
  - [x] crossChainSwap
  - [x] crossChainSwapTransfer
  - [x] crossChainSwapBTC
  - [x] crossChainSwapBTCTransfer
  - [x] withdrawZRC20
  - [x] transferBTC
  - [ ] singleChainSwap

## Cross-Chain Messaging Example

Screenshot 2023-11-20 at 12 09 34

## Cross-Chain Transaction Tracking

Screenshot 2023-11-20 at 12 27 07

### CODE REVIEW

**Comment by fadeev** [comment]
@SocketSecurity ignore next@13.5.6

**Comment by lucas-janon** [comment]
If you check the network tab, there are a lot of unneeded calls to the nodes:
image

I'm not sure about our agreement with BlockPi but making so many calls can get really expensive.
Since most of them are `eth_chainId` calls, you can get a big improvement if you use Ethers `StaticJsonRpcProvider`.

**Comment by lucas-janon** [comment]
Do we want to put guardrails in place or will that be on the user? (e.g.: preventing transfers of the same asset in the same network.)
image

**Comment by lucas-janon** [comment]
It's showing NaN on the fee, probably because convertGasToZeta is throwing a 500 error?

image

**Comment by lucas-janon** [comment]
Getting the same on BSC (`convertGasToZeta` doesn't throw on BSC).

image

**Comment by lucas-janon** [comment]
I don't know if RPCs are broken or what's happening, but I also can't send a cross-chain message:
image

**Comment by fadeev** [comment]
> I'm not sure about our agreement with BlockPi but making so many calls can get really expensive.
Since most of them are eth_chainId calls, you can get a big improvement if you use Ethers StaticJsonRpcProvider.

@lucas-janon thanks for the suggestion! I've switched to static https://github.com/zeta-chain/toolkit/pull/79/commits/408e47e6125bd0156537d9da8023045fd6bdebc4 and this reduced the number of requests by 30%. Nice!

**Comment by fadeev** [comment]
> Do we want to put guardrails in place or will that be on the user? (e.g.: preventing transfers of the same asset in the same network.)

@lucas-janon This is a feature, not a bug. The goal of this component is to allow user all possible interactions with supported tokens and a simple single chain transfer is one of them. You can change the recipient address, so it's a perfectly valid interaction, I think.

**Comment by fadeev** [comment]
> I don't know if RPCs are broken or what's happening, but I also can't send a cross-chain message:

@lucas-janon this should've been fixed in https://github.com/zeta-chain/example-frontend/pull/2/commits/e4675b8fea2ac3d45e164f731d011fe3a0aa7648. The problem was with CCM fees, you needed to know the amount of native gas tokens you need to provide to cover CCM fees denominated in ZETA. I've updated the component to fetch CCM fees and calculate the native gas tokens that you need to cover the fee. It's using Uniswap v2 on Goerli and BSC and v3 on Mumbai.

Right now, you can't send messages **from** BSC. @andresaiello is aware of this and we're looking into this. Since we can't send messages via the CLI or explorers, the problem is not related to this frontend.

**Comment by fadeev** [comment]
> Do we want to put guardrails in place or will that be on the user? (e.g.: preventing transfers of the same asset in the same network.)

@lucas-janon should be fixed in https://github.com/zeta-chain/example-frontend/pull/2/commits/cab0a50de1837e993531ca6bbf4e394926483812

Screenshot 2023-11-23 at 12 07 10

Screenshot 2023-11-23 at 12 07 22

**Comment by fadeev** [comment]
This component is still not feature-complete, though. For example, you can't swap WZETA from a connected chain to WZETA on ZetaChain. The connector automatically unwraps WZETA on ZetaChain (which is a good thing), so this will be handled on the frontend by either removing WZETA as a destination token or showing an alert.

**Comment by fadeev** [comment]
Added more states for the send button.

Screenshot 2023-11-24 at 11 00 34
Screenshot 2023-11-24 at 11 00 41
Screenshot 2023-11-24 at 11 00 47
Screenshot 2023-11-24 at 11 00 53
Screenshot 2023-11-24 at 11 00 59
Screenshot 2023-11-24 at 11 00 28

**Comment by fadeev** [comment]
@lucas-janon please, review again.

**Comment by fadeev** [comment]
Added debounce, reducing the number of API calls to one-third.

**Comment by fadeev** [comment]
Added the WIP pools page. If there is something wrong with it, I can just disable it for now, not a big deal.

**Comment by fadeev** [comment]
Disabled pools for now as they're WIP.

**Comment by lucas-janon** [comment]
https://github.com/zeta-chain/example-frontend/assets/7400784/79beaff8-3e4d-4160-bd8c-908a127c3a94

There's kind of a race condition when you change network, changing network seems to also trigger the swap, and sometimes (you can see it on the last try) the swap goes first and it asks you for allowance before changing network (or it asks you to swap if you already gave allowance).

**Comment by lucas-janon** [comment]
image

Am I supposed to set the fee myself?

**Comment by lucas-janon** [comment]
Not working for bsc, could be the RPC:
image

**Comment by lucas-janon** [comment]
Navigation detail: I'd put the cross-chain messaging button in the header instead of below "portfolio", so it's clearer that it's another page.

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
```
[the remark this answers is not in the document]
Comment (fadeev) [reply .eslintrc.json]: 33e5778

**Comment by fadeev** [comment]
> There's kind of a race condition when you change network, changing network seems to also trigger the swap, and sometimes (you can see it on the last try) the swap goes first and it asks you for allowance before changing network (or it asks you to swap if you already gave allowance).

Good catch! Should be fixed in 7bbcd8b

**Comment by fadeev** [comment]
> Am I supposed to set the fee myself?

The fee is calculated automatically and the field gets pre-filled, but you may set in manually (as a fallback if the calculation for some reason wasn't accurate). If the fee wasn't calculated when you checked it was because the RPC was very slow.

**Comment by fadeev** [comment]
I'll probably revamp the messaging form, but I want to ship something first.

**Comment by fadeev** [comment]
Added depositing supported ERC-20 tokens (currently, only USDC) through the custody contract to ZetaChain as ZRC-20.

**Review by lucas-janon — APPROVED** [review]
There are some code patterns that aren't usual in React, like having side-effects on reducers, but depending on the goal of this app I think we can let those go and merge it. Nothing else to add on top of the comments I left previously.

**Pull request merged**